Makefile

```diff
@@ -37,8 +37,20 @@ all: app escript
 erls = $(wildcard src/*.erl)
 beams = $(erls:src/%.erl=ebin/%.beam)
 
-ERLC_OPTS = -I include -I src -pa ebin +debug_info
-TEST_ERLC_OPTS = +debug_info
+define otp_version
+io:format("~s\n", [erlang:system_info(otp_release)]), halt().
+endef
+OTP_VERSION := $(shell erl -noinput -eval '$(otp_version)')
+
+ERL_OPTS 		=
+ERLC_OPTS 		= -I include -I src -pa ebin +debug_info
+TEST_ERLC_OPTS 	= +debug_info
+
+ifeq ($(shell [ $(OTP_VERSION) -ge 25 ] && echo y), y)
+ERL_OPTS 		+= -enable-feature maybe_expr
```

Comment:
This looks like Erlang.mk magic. :grin: I prefer simplicity...

Instead of passing this flag to `erlc`, I believe we can enable it using `-feature(maybe_expr).` in the erl file itself.

We can pass `-enable-feature maybe_expr` to `erl`. It seems to ignore this on older versions.

Then we don't need this makefile magic. Would it work?

Reply:
> I believe we can enable it using `-feature(maybe_expr, enable)`. in the `.erl` file itself.

Indeed, it can be skipped in the Makefile. We can't skip passing the flag to `erlc`, though.

> Then we don't need this makefile magic. Would it work?

We still need it to control that pre-25 we don't want to compile the test file in the extra path: `test/otp25/known_problems/should_pass/maybe_expr.erl`. Otherwise, because there's only one test in the `maybe_expr.erl` file that's feature gated, on older versions checking the file passes, therefore it's raised that the file shouldn't be a known problem. See CI logs: https://github.com/josefs/Gradualizer/actions/runs/3576727016/jobs/6014886961

We either need the condition on OTP version in the Makefile or we need to inject an `else` preprocessor clause and a bogus (failing) test to `maybe_expr.erl` itself, so it stays a known problem no matter which branch the preprocessor takes.

Reply:
For the `should_fail` tests, I think the test framework needs to be changed to detect passing functions within a file of known problems. I think it should check that there is one error for each exported function, so there's 0 errors for 0 functions, that's "all functions fail".